File: android/src/main/java/com/rajatmalhotra/plugins/App_n_DeviceInfoPlugin.java
package com.rajatmalhotra.plugins;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.Locale;

@CapacitorPlugin(name = "App_n_DeviceInfo")
public class App_n_DeviceInfoPlugin extends Plugin {

    private App_n_DeviceInfo implementation;

    @Override
    public void load() {
        implementation = new App_n_DeviceInfo(getContext());
    }

    @PluginMethod
    public void getId(PluginCall call) {
        JSObject r = new JSObject();

        r.put("uuid", implementation.getUuid());

        call.resolve(r);
    }

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void getDeviceInfo(PluginCall call) {
        JSObject r = new JSObject();

        r.put("memUsed", implementation.getMemUsed());
        r.put("diskFree", implementation.getDiskFree());
        r.put("diskTotal", implementation.getDiskTotal());
        r.put("realDiskFree", implementation.getRealDiskFree());
        r.put("realDiskTotal", implementation.getRealDiskTotal());
        r.put("model", android.os.Build.MODEL);
        r.put("operatingSystem", "android");
        r.put("osVersion", android.os.Build.VERSION.RELEASE);
        r.put("platform", implementation.getPlatform());
        r.put("manufacturer", android.os.Build.MANUFACTURER);
        r.put("isVirtual", implementation.isVirtual());
        r.put("name", implementation.getName());
        r.put("webViewVersion", implementation.getWebViewVersion());

        call.resolve(r);
    }

    @PluginMethod
    public void getAppInfo(PluginCall call) {
        JSObject r = new JSObject();
        PackageManager manager = getActivity().getPackageManager();
        PackageInfo info = null;
        try {
            info = manager.getPackageInfo(getActivity().getPackageName(), PackageManager.GET_ACTIVITIES);
            r.put("version", info.versionName);
            r.put("build", info.versionCode);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        call.resolve(r);
    }

    @PluginMethod
    public void getLanguageCode(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("value", Locale.getDefault().getLanguage());
        call.resolve(ret);
    }

    @PluginMethod
    public void getBatteryInfo(PluginCall call) {
        JSObject r = new JSObject();

        r.put("batteryLevel", implementation.getBatteryLevel());
        r.put("isCharging", implementation.isCharging());

        call.resolve(r);
    }
}

File: ios/Plugin/App_n_DeviceInfoPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(App_n_DeviceInfoPlugin)
public class App_n_DeviceInfoPlugin: CAPPlugin {
    private let implementation = App_n_DeviceInfo()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    @objc func getDeviceInfo(_ call: CAPPluginCall) {
        var isSimulator = false
        #if targetEnvironment(simulator)
        isSimulator = true
        #endif

        let memUsed = implementation.getMemoryUsage()
        let diskFree = implementation.getFreeDiskSize() ?? 0
        let realDiskFree = implementation.getRealFreeDiskSize() ?? 0
        let diskTotal = implementation.getTotalDiskSize() ?? 0
        let currentDeviceModel = implementation.getCurrentDeviceModel()

        call.resolve([
            "memUsed": memUsed,
            "diskFree": diskFree,
            "diskTotal": diskTotal,
            "realDiskFree": realDiskFree,
            "realDiskTotal": diskTotal,
            "name": UIDevice.current.name,
            "model": UIDevice.current.model,
            "localModel": UIDevice.current.localizedModel,
            "currentDeviceModel": currentDeviceModel,
            "operatingSystem": "ios",
            "osVersion": UIDevice.current.systemVersion,
            "platform": "ios",
            "manufacturer": "Apple",
            "isVirtual": isSimulator,
            "webViewVersion": UIDevice.current.systemVersion
        ])
    }

    @objc func getLanguageCode(_ call: CAPPluginCall) {
        let code = implementation.getLanguageCode()
        call.resolve([
            "value": code
        ])
    }
    @objc func getBatteryInfo(_ call: CAPPluginCall) {
        UIDevice.current.isBatteryMonitoringEnabled = true

        call.resolve([
            "batteryLevel": UIDevice.current.batteryLevel,
            "isCharging": UIDevice.current.batteryState == .charging || UIDevice.current.batteryState == .full
        ])

        UIDevice.current.isBatteryMonitoringEnabled = false
    }
    @objc func getAppInfo(_ call: CAPPluginCall) {
        let currentVersion = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as! String
        let currentBuild = Bundle.main.object(forInfoDictionaryKey: kCFBundleVersionKey as String) as! String
        call.resolve([
            "version": currentVersion,
            "build": currentBuild
        ])
    }    
}
